src/shims/unix/linux/epoll.rs

```diff
@@ -500,10 +583,23 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
                         let event_instance =
                             EpollEventInstance::new(flags, epoll_event_interest.data);
                         ready_list.insert(epoll_key, event_instance);
+                        // Edge-triggered notification only notify one thread even if there are
+                        // multiple threads block on the same epfd.
+                        let epfd = this.machine.fds.get(epoll_event_interest.epfd).unwrap();
+                        // FIXME: We can randomly pick a thread to unblock.
+                        if let Some(thread_id) =
+                            epfd.downcast::<Epoll>().unwrap().thread_id.borrow_mut().pop()
+                        {
+                            waiter.push(thread_id);
+                        };
                     }
                 }
             }
         }
+        waiter.dedup();
```

Comment:
You need to sort first. Dedup only dedups consecutive occurrences of the same value

Reply:
Ah I see. I misread the documentation.